moshi/src/main/java/com/squareup/moshi/StandardJsonAdapters.java

```diff
@@ -212,27 +213,46 @@ static int rangeCheckNextInt(JsonReader reader, String typeMessage, int min, int
     }
   };
 
-  static <T extends Enum<T>> JsonAdapter<T> enumAdapter(final Class<T> enumType) {
-    return new JsonAdapter<T>() {
-      @Override public T fromJson(JsonReader reader) throws IOException {
-        String name = reader.nextString();
-        try {
-          return Enum.valueOf(enumType, name);
-        } catch (IllegalArgumentException e) {
-          throw new JsonDataException("Expected one of "
-              + Arrays.toString(enumType.getEnumConstants()) + " but was " + name + " at path "
-              + reader.getPath());
+  static final class EnumJsonAdapter<T extends Enum<T>> extends JsonAdapter<T> {
+    private final Class<T> enumType;
+    private final Map<String, T> nameConstantMap = new LinkedHashMap<>();
+
+    public EnumJsonAdapter(Class<T> enumType) {
+      this.enumType = enumType;
+      try {
+        T[] constants = enumType.getEnumConstants();
+        for (T constant : constants) {
+          Json annotation = enumType.getField(constant.name()).getAnnotation(Json.class);
+          String name = annotation != null ? annotation.name() : constant.name();
+          nameConstantMap.put(name, constant);
         }
+      } catch (NoSuchFieldException e) {
+        throw new AssertionError("Missing field in " + enumType.getName(), e);
```

Comment:
btw, fi you going to follow up on this pr... consider changing this to `throw new AssertionError("Missing field in " + enumType.getName());`. The `AssertionError(String, Throwable)` constructor is available on Android only since KitKat.